File: Queues/ImplementQueueLinkedList.java
/* File Name: QueueLinkedList.java */

class Node
{
    int data;
    Node next;

    Node(int data)
    {
        this.data = data;
        this.next = null;
    }
}

class QueueLinkedList
{
    Node front = null, rear = null;

    void enqueue(int item)
    {
        Node newNode = new Node(item);
        if (rear == null)
            front = rear = newNode;
        else
        {
            rear.next = newNode;
            rear = newNode;
        }
    }

    void dequeue()
    {
        if (front == null)
        {
            System.out.println("Queue Underflow");
            return;
        }
        System.out.println("Deleted " + front.data);
        front = front.next;
        if (front == null)
            rear = null;
    }

    void display()
    {
        if (front == null)
        {
            System.out.println("Queue is Empty");
            return;
        }
        Node temp = front;
        while (temp != null)
        {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    public static void main(String[] args)
    {
        QueueLinkedList q = new QueueLinkedList();
        q.enqueue(10);
        q.enqueue(20);
        q.enqueue(30);
        q.display();
        q.dequeue();
        q.display();
    }
}

File: Queues/ImplementQueueLinkedList.c
/* File Name: QueueLinkedList.c */
#include <stdio.h>
#include <stdlib.h>

struct Node
{
    int data;
    struct Node *next;
};

struct Node *front = NULL, *rear = NULL;

void enqueue(int item)
{
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    newNode->data = item;
    newNode->next = NULL;

    if (rear == NULL)
        front = rear = newNode;
    else
    {
        rear->next = newNode;
        rear = newNode;
    }
}

void dequeue()
{
    if (front == NULL)
    {
        printf("Queue Underflow\n");
        return;
    }
    struct Node *temp = front;
    printf("Deleted %d\n", temp->data);
    front = front->next;
    if (front == NULL)
        rear = NULL;
    free(temp);
}

void display()
{
    struct Node *temp = front;
    if (temp == NULL)
    {
        printf("Queue is Empty\n");
        return;
    }
    while (temp != NULL)
    {
        printf("%d ", temp->data);
        temp = temp->next;
    }
    printf("\n");
}

int main()
{
    enqueue(10);
    enqueue(20);
    enqueue(30);
    display();
    dequeue();
    display();
    return 0;
}
